### PR TITLE
firing/hooking takes aim when action happened not when tick

### PR DESCRIPTION
**Pull request by JSaurusRex**
When moving the mouse fast there can be a discrepancy between where I aimed and where my weapon or hook actually go's. This is because of the fact that the mouse can move after I've shot/hooked and the mouse position at the end of the tick gets send, not the one when shot/hooked.
This makes aiming these actions kind of random and tick dependent, if you did the action right before the tick ends its really accurate but if its all the way in the beginning the mouse can move up to 20 ms longer, that can be a big difference when moving fast.
Hopefully tying it to framerate will behave better, since most players play with higher framerates than 50.

That all being said, I have tested this but there isn't an extremely noticeable difference, some pro-players might want to test this out. That being said, I have printed the difference in values, the highest I saw was 100x 150y at 1920x1080. That probably is a pretty decent difference in aim, can be the difference between hitting a shot or hook and not.

Also sorry for shitty code I don't know the coding style that well and just wanted to quickly implement it, I don't mind changing the implementation.

## Checklist

- [X] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [x] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [x] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Review by def- — CHANGES_REQUESTED** [review]
I think this is by design, it might also make some previous tricks impossible, so I'm strongly opposed.

**Comment by JSaurusRex** [comment]
It 100% should be an option that can be disabled / enabled. However I think it will also make a lot of tricks easier, I cant think of any tricks it would make harder. "I think this is by design", what do you mean by this? With my PR it still is 50hz nor does it do anything out of order.

**Comment by def-** [comment]
I think these intentionally act immediately and no on the next tick. You are changing the physics of the game, so this would break any rank.

We don't want further options. That would increase complexity.

**Comment by JSaurusRex** [comment]
At larger speeds this would pr would actually end up hurting the accuracy compared to what the player really wanted to do, I can implement a if statement to switch between using this and not depending whichever is greater but at that point it really has gotten more drawbacks than might be worth.
Still worth it for some modes (gores and fng for example, low movement speed but need fast mouse movements), but definitely not an 1 fix fixes all solution. I'll think about what to change next (probably opt in setting)

**Comment by JSaurusRex** [comment]
> I think these intentionally act immediately and no on the next tick. You are changing the physics of the game, so this would break any rank.
> 
> We don't want further options. That would increase complexity.

no worries this doesn't act immediately, it simply replaces the position of the mouse that gets send with the position of when shot, so it doesn't change physics at all (which also is impossible from the client side alone)

**Comment by trml** [comment]
From what I understand this pr attempts solves a problem that goes something like this:
You press fire when you are aiming at mousepos X, but there is still some time left (up to 20 ms) until the client will send the next input. Once the next tick is reached, the client sends mousepos Y instead of X, because the mouse moved a little in between.

The goal of this pr is to send mousepos X instead, which removes the randomness of which mouse position that is sent (since now it is random from the users point of view), and also on average reduces input lag by 10 milliseconds.

To me this sounds like a good idea (I'm not sure if a config is needed? reducing input lag is always a good thing?)

**Comment by JSaurusRex** [comment]
> From what I understand this pr attempts solves a problem that goes something like this: You press fire when you are aiming at mousepos X, but there is still some time left (up to 20 ms) until the client will send the next input. Once the next tick is reached, the client sends mousepos Y instead of X, because the mouse moved a little in between.
> 
> The goal of this pr is to send mousepos X instead, which removes the randomness of which mouse position that is sent (since now it is random from the users point of view), and also on average reduces input lag by 10 milliseconds.
> 
> To me this sounds like a good idea (I'm not sure if a config is needed? reducing input lag is always a good thing?)

yes this exactly what it does, it doesn't really reduce input lag since its still just as delayed when you or the server see it but it kinda acts similarly in functionality. Only thing is that it can be less accurate when moving fast, imagine you're moving fast and tracking an tile with your mouse to hook, you hook but because the character moves up to 20ms further you miss. This is why it should be decided by the user and it depends per gamemode and personal preference if the player prefers it, if you track your hooks / shots it will decrease accuracy but if you flick it will increase. Some logic could be implemented but I think having it as an option (by default off) is the way to go. Some future PR can implement more complex logic if they wanted to.

**Comment by trml** [comment]
> if you track your hooks / shots it will decrease accuracy but if you flick it will increase

I didn't think about tracking your shots, but yes, in that case having it enabled could create some inconsistencies (some logic could perhaps be added to handle the moving player case, but not for the moving target case).

File: src/game/client/components/binds.cpp [hunk 1]
```diff
@@ -134,10 +135,15 @@ bool CBinds::OnInput(const IInput::CEvent &Event)
 	Mask &= ~KeyModifierMask;
 
 	bool ret = false;
+	char *key = 0;
```
Comment (def-) [inline src/game/client/components/binds.cpp:138]: pKey

File: src/game/client/components/binds.cpp [hunk 2]
```diff
@@ -1,6 +1,7 @@
 /* (c) Magnus Auvinen. See licence.txt in the root of the distribution for more information. */
 /* If you are missing that file, acquire a complete release at teeworlds.com.                */
 #include "binds.h"
+#include <cstring>
```
Comment (def-) [inline src/game/client/components/binds.cpp:4]: We don't use cstring directly, only system.h

File: src/game/client/components/binds.cpp [hunk 3]
```diff
@@ -148,13 +154,24 @@ bool CBinds::OnInput(const IInput::CEvent &Event)
...
+		}
 		// Have to check for nullptr again because the previous execute can unbind itself
 		if(Event.m_Flags & IInput::FLAG_RELEASE && m_aapKeyBindings[0][Event.m_Key])
 			Console()->ExecuteLineStroked(0, m_aapKeyBindings[0][Event.m_Key]);
 		ret = true;
 	}
 
+	if(g_Config.m_ClSubTickAiming && key)
+	{
+		if(strcmp("+fire", key) == 0 || strcmp("+hook", key) == 0)
```
Comment (def-) [inline src/game/client/components/binds.cpp:169]: str_comp from system.h

File: src/game/client/components/binds.h [hunk 4]
```diff
@@ -35,6 +35,8 @@ class CBinds : public CComponent
 		virtual bool OnInput(const IInput::CEvent &Event) override;
 	};
 
+	bool mouseOnAction;
```
Comment (def-) [inline src/game/client/components/binds.h:38]: MouseOnAction

**Review by def- — CHANGES_REQUESTED** [review]
I'm still afraid of this change, but since it's off by default let's give it a try

File: src/game/client/gameclient.cpp [hunk 5]
```diff
@@ -410,6 +410,12 @@ void CGameClient::OnUpdate()
 				break;
 		}
 	}
+
+	if(g_Config.m_ClSubTickAiming && m_Binds.mouseOnAction)
+	{
+		m_Controls.m_aMousePosOnAction[g_Config.m_ClDummy] = m_Controls.m_aMousePos[g_Config.m_ClDummy]; //idk what g_Config.m_ClDummy is
```
Comment (Robyt3) [inline src/game/client/gameclient.cpp:416]: Remove the comment. It means whether dummy is currently active or not.

File: src/game/client/components/binds.cpp (hunk 1 above)
Comment (Robyt3) [reply src/game/client/components/binds.cpp:138]: And `const char *` instead of `char *`

File: src/game/client/components/binds.h (hunk 4 above)
Comment (Robyt3) [reply src/game/client/components/binds.h:38]: `m_MouseOnAction`

File: src/game/client/components/binds.cpp (hunk 1 above)
Comment (Robyt3) [reply src/game/client/components/binds.cpp:138]: `nullptr` instead of `0`

File: src/game/client/gameclient.cpp (hunk 5 above)
Comment (JSaurusRex) [reply src/game/client/gameclient.cpp:416]: forgot I added that xD, will do

**Review by def- — APPROVED** [review]

**Pull request merged**